File: web/src/noise/frame.js
// web/src/noise/frame.js
// Mirrors go/internal/noise/frame.go byte-for-byte.
export const FRAME_DATA = 0x01;
export const FRAME_RESIZE = 0x02;
export const FRAME_HELLO = 0x03;

export function encodeData(bytes) {
  const out = new Uint8Array(1 + bytes.length);
  out[0] = FRAME_DATA;
  out.set(bytes, 1);
  return out;
}

export function encodeResize(cols, rows) {
  const out = new Uint8Array(5);
  out[0] = FRAME_RESIZE;
  const dv = new DataView(out.buffer);
  dv.setUint16(1, cols, false); // big-endian
  dv.setUint16(3, rows, false);
  return out;
}

export function encodeHello(jsonBytes) {
  const out = new Uint8Array(1 + jsonBytes.length);
  out[0] = FRAME_HELLO;
  out.set(jsonBytes, 1);
  return out;
}

export function decodeFrame(bytes) {
  if (bytes.length < 1) throw new Error('empty frame');
  return { type: bytes[0], payload: bytes.slice(1) };
}

export function decodeResize(payload) {
  if (payload.length !== 4) throw new Error('resize payload must be 4 bytes');
  const dv = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
  return { cols: dv.getUint16(0, false), rows: dv.getUint16(2, false) };
}

File: go/internal/noise/frame.go
// go/internal/noise/frame.go
package noise

import (
	"encoding/binary"
	"errors"
)

// FrameType is the 1-byte tag prefixing every frame inside the Noise channel.
type FrameType byte

const (
	FrameData    FrameType = 0x01 // raw PTY bytes
	FrameResize  FrameType = 0x02 // cols u16 BE ++ rows u16 BE
	FrameHello   FrameType = 0x03 // UTF-8 JSON metadata
	FrameWindows FrameType = 0x04 // UTF-8 JSON tmux window snapshot (agent -> client)
)

// EncodeData wraps raw PTY bytes in a DATA frame.
func EncodeData(b []byte) []byte {
	return append([]byte{byte(FrameData)}, b...)
}

// EncodeResize encodes terminal dimensions as a RESIZE frame.
func EncodeResize(cols, rows uint16) []byte {
	out := make([]byte, 5)
	out[0] = byte(FrameResize)
	binary.BigEndian.PutUint16(out[1:3], cols)
	binary.BigEndian.PutUint16(out[3:5], rows)
	return out
}

// EncodeHello wraps JSON metadata in a HELLO frame.
func EncodeHello(jsonBytes []byte) []byte {
	return append([]byte{byte(FrameHello)}, jsonBytes...)
}

// EncodeWindows wraps a tmux window snapshot (JSON) in a WINDOWS frame.
func EncodeWindows(jsonBytes []byte) []byte {
	return append([]byte{byte(FrameWindows)}, jsonBytes...)
}

// DecodeFrame splits a frame into its type and payload.
func DecodeFrame(b []byte) (FrameType, []byte, error) {
	if len(b) < 1 {
		return 0, nil, errors.New("empty frame")
	}
	return FrameType(b[0]), b[1:], nil
}

// DecodeResize parses a RESIZE payload into cols and rows.
func DecodeResize(payload []byte) (cols, rows uint16, err error) {
	if len(payload) != 4 {
		return 0, 0, errors.New("resize payload must be 4 bytes")
	}
	return binary.BigEndian.Uint16(payload[0:2]), binary.BigEndian.Uint16(payload[2:4]), nil
}
